processing/src/main/java/io/druid/query/dimension/LookupDimensionSpec.java

```diff
@@ -0,0 +1,251 @@
+
+  @JsonProperty
+  private final LookupExtractor lookup;
+
+  @JsonProperty
+  private final boolean retainMissingValues;
+
+  @JsonProperty
+  private final String replaceMissingWith;
+
+  @JsonProperty
+  private final String name;
+
+
+  private final LookupReferencesManager lookupReferencesManager;
+
+
+  @JsonCreator
+  public LookupDimensionSpec(
+      @JsonProperty("dimension") String dimension,
+      @JsonProperty("outputName") String outputName,
+      @JsonProperty("lookup") LookupExtractor lookup,
+      @JsonProperty("retainMissingValues") final boolean retainMissingValues,
+      @JsonProperty("replaceMissingWith") final String replaceMissingWith,
+      @JsonProperty("name") String name,
+      @JacksonInject LookupReferencesManager lookupReferencesManager
+  )
+  {
+    this.retainMissingValues = retainMissingValues;
+    this.replaceMissingWith = Strings.emptyToNull(replaceMissingWith);
+    Preconditions.checkArgument(
+        !(retainMissingValues && !Strings.isNullOrEmpty(replaceMissingWith)),
+        "Cannot specify a [replaceMissingValue] and set [retainMissingValue] to true"
+    );
+    this.dimension = Preconditions.checkNotNull(dimension, "dimension can not be Null");
+    this.outputName = Preconditions.checkNotNull(outputName, "outputName can not be Null");
+    this.lookupReferencesManager = lookupReferencesManager;
+    this.name = name;
+    this.lookup = lookup;
+    Preconditions.checkArgument(
+        Strings.isNullOrEmpty(name) ^ (lookup == null),
+        "name and lookup are mutually exclusive please provide either a name or a lookup"
+    );
+  }
+
+  @Override
+  @JsonProperty
+  public String getDimension()
+  {
+    return dimension;
+  }
+
+  @Override
+  @JsonProperty
+  public String getOutputName()
+  {
+    return outputName;
+  }
+
+  @JsonProperty
+  @Nullable
+  public LookupExtractor getLookup()
+  {
+    return lookup;
+  }
+
+  @JsonProperty
+  @Nullable
+  public String getName()
+  {
+    return name;
+  }
+
+  @Override
+  public ExtractionFn getExtractionFn()
+  {
+    final LookupExtractor lookupExtractor = Strings.isNullOrEmpty(name) ? this.lookup
+                                                                        : Preconditions.checkNotNull(
+                                                                            Preconditions.checkNotNull(
+                                                                                this.lookupReferencesManager,
+                                                                                "can not find lookup manager"
+                                                                            ).get(name).get(),
+                                                                            "can not find lookup with name [%s]",
+                                                                            name
+                                                                        );
+    return new DimExtractionFn()
```

Comment:
it would look something like this:

```java
new FunctionalExtraction(
    new Function<String, String>(){
          @Nullable
          @Override
          public String apply(String input)
          {
            return lookupExtractor.apply(Strings.nullToEmpty(input));
          }
    },
    retainMissingValue,
    replaceMissingValueWith,
    lookupExtractor.isOneToOne()
)
```